src/view.c

```diff
@@ -874,6 +875,69 @@ view_center(struct view *view, const struct wlr_box *ref)
 	}
 }
 
+/*
+ * Algorithm copied from KWin
+ * https://github.com/KDE/kwin/blob/df9f8f8346b5b7645578e37365dabb1a7b02ca5a/src/placement.cpp#L589
+ */
+static void
+view_cascade(struct view *view)
+{
+	/* Default geometry to which the view is moved */
+	struct wlr_box center = view->pending;
+	view_compute_centered_position(view, NULL,
+		center.width, center.height, &center.x, &center.y);
+
+	/* SSD-aware candidate geometry to which the view is moved */
+	struct wlr_box self = center;
+	struct border margin = ssd_get_margin(view->ssd);
+	self.x -= margin.left;
+	self.y -= margin.top;
+	self.width += margin.left + margin.right;
+	self.height += margin.top + margin.bottom;
+
+	struct wlr_box usable = output_usable_area_in_layout_coords(view->output);
+
+cascade_if_covers:;
+	struct wlr_box covered = {0};
+
+	/* Iterate over views from top to bottom */
+	struct view *other_view;
+	for_each_view(other_view, &view->server->views,
+			LAB_VIEW_CRITERIA_CURRENT_WORKSPACE) {
+		struct wlr_box other = ssd_max_extents(other_view);
+		if (other_view == view || !wlr_box_intersects(&self, &other)) {
+			continue;
+		}
+		/*
+		 * If the candidate geometry covers an existing view which is not
+		 * already covered, shift the candidate geometry to bottom-right.
+		 */
+		if (wlr_box_contains(&self, &other)
+				&& !wlr_box_contains(&covered, &other)) {
+			self.x = other.x + rc.placement_cascade_offset_x;
+			self.y = other.y + rc.placement_cascade_offset_y;
+			/*
+			 * Fall back to center if the canditate geometry
+			 * doesn't fit within the usable area.
+			 */
+			if (!wlr_box_contains(&usable, &self)) {
+				view_move(view, center.x, center.y);
+				return;
+			}
+			/* Repeat with the new candidate geometry */
+			goto cascade_if_covers;
```

Comment:
I used `goto` for simplicity, but KWin uses a flag variable `noOverlap` + `break;` instead ([link](https://github.com/KDE/kwin/blob/df9f8f8346b5b7645578e37365dabb1a7b02ca5a/src/placement.cpp#L613-L620)).

Perhaps we should only use `goto` for cleanup logic and avoid using it for loop.

Reply:
I'm normally quite relaxed about `goto`s, but they are often derided by more learned people (than me).

I think `kwin`'s `while (!noOverlap)` certainly makes it easier for the reader to understand what's going on. So on the basis that readability is important, I think we ought to wrap it in a while-loop.

Reply:
Fixed.